### PR TITLE
Fixed redefination warning messages for _GNU_SOURCE

### PR DESCRIPTION
**Pull request by nileshkale123**
## Description

This commit solves the issue of redefinition warning messages for _GNU_SOURCE while building on Linux platform.

## PR checklist

Please tick as appropriate and edit the reasons (e.g.: "backport: not needed because this is a new feature")

- [x] **changelog** provided
- [x] **2.28 backport** #9086
- [x] **3.6 backport** #9085
- [x] **tests** not required

## Notes for the submitter

Please refer to the [contributing guidelines](https://github.com/Mbed-TLS/mbedtls/blob/development/CONTRIBUTING.md), especially the
checklist for PR contributors.

Help make review efficient:
* Multiple simple commits
  - please structure your PR into a series of small commits, each of which does one thing
* Avoid force-push
  - please do not force-push to update your PR - just add new commit(s)
* See our [Guidelines for Contributors](https://mbed-tls.readthedocs.io/en/latest/reviews/review-for-contributors/) for more details about the review process.

### CODE REVIEW

File: ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
```
Comment (yanesca) [inline ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt:1]: We are processing these files with a script and it needs the first line to be non-empty. Could you please delete this line.

**Review by yanesca — CHANGES_REQUESTED** [review]
Hi @nileshkale123, thank you for your contribution.

After fixing the superfluous empty line, could you please raise backport PRs against the `mbedtls-2.28` and `mbedtls-3.6` branches?

File: ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Bugfix
+   * Fix issue of redefination warning messages for _GNU_SOURCE in 
+     entropy_poll.c and sha_256.c. There was build warnng during
```
Comment (yanesca) [inline ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt:3]:
```suggestion
     entropy_poll.c and sha_256.c. There was build warning during
```

**Review by yanesca — CHANGES_REQUESTED** [review]
There is a newline character missing from the end of the last line which makes the CI fail. Could you please fix that?

**Review by yanesca — DISMISSED** [review]
LGTM

File: ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt [hunk 3]
```diff
@@ -0,0 +1,5 @@
+Bugfix
+   * Fix issue of redefination warning messages for _GNU_SOURCE in
```
Comment (tom-daubney-arm) [inline ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt:2]:
```suggestion
   * Fix issue of redefinition warning messages for _GNU_SOURCE in
```

File: ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+     entropy_poll.c and sha_256.c. There was build warning during
```
Comment (tom-daubney-arm) [inline ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt:3]:
```suggestion
     entropy_poll.c and sha_256.c. There was a build warning during
```

**Review by tom-daubney-arm — CHANGES_REQUESTED** [review]
Looks good except for a couple of small typos.

**Review by yanesca — APPROVED** [review]
LGTM

**Review by tom-cosgrove-arm — CHANGES_REQUESTED** [review]
The CI failure is relevant:

```
[2024-04-30T08:35:15.580Z] ******************************************************************
[2024-04-30T08:35:15.580Z] * check_files: Check: file sanity checks (permissions, encodings) 
[2024-04-30T08:35:15.580Z] * Tue Apr 30 08:35:15 UTC 2024
[2024-04-30T08:35:15.580Z] ******************************************************************
[2024-04-30T08:35:22.185Z] Trailing whitespace:
[2024-04-30T08:35:22.185Z] ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt: 2, 3
[2024-04-30T08:35:22.185Z] 
[2024-04-30T08:35:22.185Z] ^^^^check_files: Check: file sanity checks (permissions, encodings): tests/scripts/check_files.py -> 1^^^^
```

Please remove the trailing whitespace in `ChangeLog.d/fix-redefination_warning_messages_for_GNU_SOURCE.txt`

**Review by tom-cosgrove-arm — APPROVED** [review]
Backports to 3.6 and 2.28 are still required

**Comment by tom-cosgrove-arm** [comment]
@nileshkale123 Are you able to create the backports for the `mbedtls-2.28` and `mbedtls-3.6` branches?

**Comment by nileshkale123** [comment]
> @nileshkale123 Are you able to create the backports for the `mbedtls-2.28` and `mbedtls-3.6` branches?

Yes, I have created MR with issue ID 9085 & 9086

**Pull request merged**